core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go

```diff
@@ -229,43 +234,51 @@ func (provider *GitPackageContentProvider) GetAbsoluteLocatorForRelativeLocator(
 
-	found, packageToBeReplaced, replaceWithPackage := findPackageReplace(absoluteLocator, packageReplaceOptions)
-	if found {
-		replacedAbsoluteLocator := strings.Replace(absoluteLocator, packageToBeReplaced, replaceWithPackage, onlyOneReplace)
-		logrus.Debugf("absoluteLocator '%s' replaced with '%s'", absoluteLocator, replacedAbsoluteLocator)
-		return replacedAbsoluteLocator
+	historicalPackageReplaceOptions, err := provider.packageReplaceOptionsRepository.Get()
+	if err != nil {
+		return startosis_errors.WrapWithInterpretationError(err, "An error occurred getting the historical package replace options from the repository")
 	}
 
-	return absoluteLocator
-}
+	for packageId, historicalReplace := range historicalPackageReplaceOptions {
 
-func findPackageReplace(absoluteLocator string, packageReplaceOptions map[string]string) (bool, string, string) {
-	pathToAnalyze := absoluteLocator
-	for {
-		numberSlashes := strings.Count(pathToAnalyze, urlPathSeparator)
+		shouldClonePackage := false
 
-		// check for the minimal path e.g.: github.com/org/package
-		if numberSlashes < minimumSubPathsForValidGitURL {
-			break
+		isHistoricalLocalReplace := isLocalDependencyReplace(historicalReplace)
+		logrus.Debugf("historicalReplace '%v' isHistoricalLocalReplace? '%v', ", historicalReplace, isHistoricalLocalReplace)
+
+		currentReplace, isCurrentReplace := currentPackageReplaceOptions[packageId]
+		if isCurrentReplace {
+			// the package will be cloned if the current replace is remote and the historical is local
+			isCurrentRemoteReplace := !isLocalDependencyReplace(currentReplace)
+			logrus.Debugf("currentReplace '%v' isCurrentRemoteReplace? '%v', ", isCurrentRemoteReplace, currentReplace)
+			if isCurrentRemoteReplace && isHistoricalLocalReplace {
+				shouldClonePackage = true
+			}
 		}
-		lastIndex := strings.LastIndex(pathToAnalyze, urlPathSeparator)
 
-		packageToBeReplaced := pathToAnalyze[:lastIndex]
-		replaceWithPackage, ok := packageReplaceOptions[packageToBeReplaced]
-		if ok {
-			logrus.Debugf("dependency replace found for '%s', package '%s' will be replaced with '%s'", absoluteLocator, packageToBeReplaced, replaceWithPackage)
-			return true, packageToBeReplaced, replaceWithPackage
+		// there is no current replace for this dependency but the version in the cache is local
+		if !isCurrentReplace && isHistoricalLocalReplace {
+			shouldClonePackage = true
 		}
 
-		pathToAnalyze = packageToBeReplaced
+		if shouldClonePackage {
+			if _, err := provider.ClonePackage(packageId); err != nil {
+				return startosis_errors.WrapWithInterpretationError(err, "An error occurred cloning package '%v'", packageId)
+			}
+		}
 	}
 
-	return false, "", ""
+	// upgrade the historical-replace list with the new values
+	for packageId, currentReplace := range currentPackageReplaceOptions {
+		historicalPackageReplaceOptions[packageId] = currentReplace
```

Comment:
we call it "historical" but its actually `existingPackageRelaceOptions`; Historical sounds like if you do 5 different runs we have all 5 different (or same) values

Reply:
you are right! great suggestion, I'll do the change